Update paper.md (#70)
JuliaSmoothOptimizers/KnetNLPModels.jl: paper/paper.md

## FluxNLPModels.jl

We assume that the MNIST dataset [@lecun-bouttou-bengio-haffner1998] from MLDataset has been downloaded, loaded, and a minibatch loader has been created. Additionally, we presume that a LeNet model is defined in Flux.jl. To view an example, please refer to the 'example' folder of FluxNLPModels.jl or check out this [link](https://github.com/Farhad-phd/FluxNLPModels.jl/blob/main/example/MNIST_cnn.jl). <!-- TODO change the link to JSO ? -->

To cast the LeNet model as an FluxNLPModel, one needs to pass the model that was defined in Flux, the loss function, as well as the train and test data loaders.
Flux.jl allows flexibility to define any loss function we need.
We will use the built-in `Flux.logitcrossentropy`.

```julia
using FluxNLPModels

LeNetNLPModel = FluxNLPModel(LeNet, train_loader, test_loader; loss_f = Flux.logitcrossentropy)
```

> After completing the necessary steps, one can utilize a solver from JSOSolvers to minimize the loss of LeNetNLPModel. These solvers have been primarily designed for deterministic optimization. In the case of FluxNLPModel.jl (and KnetNLPModels.jl), the loss function is managed to ensure its application to sampled data. However, it is essential to modify the training minibatch between iterations. This can be accomplished by leveraging the callback mechanism incorporated in JSOSolvers. This mechanism executes a pre-defined callback at the conclusion of each iteration. For more comprehensive information, we refere the readers to the JSOSolvers documentation.

Reviewer: ```suggestion
After completing the necessary steps, one can utilize a solver from JSOSolvers to minimize the loss of LeNetNLPModel.
These solvers have been primarily designed for deterministic optimization. In the case of FluxNLPModel.jl (and KnetNLPModels.jl), the loss function is managed to ensure its application to sampled data.
However, it is essential to modify the training minibatch between iterations.
This can be accomplished by leveraging the callback mechanism incorporated in JSOSolvers.
This mechanism executes a pre-defined callback at the conclusion of each iteration.
For more comprehensive information, we refere the reader to the JSOSolvers documentation.
```

max_time = 300. # run at most 5min

> callback = (LeNetNLPModel, solver, stats) -> FluxNLPModels.minibatch_next_train!(LeNetNLPModel)

Reviewer: ```suggestion
callback = (LeNetNLPModel, 
                   solver, 
                   stats) -> FluxNLPModels.minibatch_next_train!(LeNetNLPModel)
```

solver_stats = R2(LeNetNLPModel; callback, max_time) # We collect the status of Solver run
test_accuracy = FluxNLPModels.accuracy(LeNetNLPModel)
```

Another choice to train `LeNetNLPModel` is the LBFGS solver with linesearch:
```julia
solver_stats = lbfgs(LeNetNLPModel; callback, max_time)
```

> To exploit any non-convexity present in `LeNetNLPModel`, an `LSR1` approximation of the Hessian which can be employed and fed into the `trunk` solver, utilizes a trust-region method with a backtracking linesearch.

Reviewer: ```suggestion
To exploit any non-convexity present in `LeNetNLPModel`, an `LSR1` approximation of the Hessian which can be employed and fed into the `trunk` solver, which utilizes a trust-region method with a backtracking linesearch.
```

solver_stats = lbfgs(LeNetNLPModel; callback, max_time)
test_accuracy = FluxNLPModels.accuracy(LeNetNLPModel)
```

To exploit any non-convexity present in `LeNetNLPModel`, an `LSR1` approximation of the Hessian can be employed and fed into the `trunk` solver.
The trunk solver utilizes a trust-region method with a backtracking line search.
To exploit any non-convexity present in `LeNetNLPModel`, an `LSR1` approximation of the Hessian which can be employed and fed into the `trunk` solver, utilizes a trust-region method with a backtracking linesearch.
To integrate the LSR1 approximation and trunk into the training process, the code can be modified as:

```julia
using NLPModelsModifiers # define also LSR1Model
using NLPModelsModifiers # defines LSR1Model

lsr1_LeNet = NLPModelsModifiers.LSR1Model(LeNetNLPModel)
callback_lsr1 = 
  (lsr1_LeNet, solver, stats) -> FluxNLPModels.minibatch_next_train!(
                                                 lsr1_LeNet.model
                                               )

> 

Reviewer: ```suggestion
solver_stats = trunk(lsr1_LeNet; callback = callback_lsr1, max_time)
```

Author: I do this manually, git gives me error

Reviewer: What is the error?
It is related to the software engineer part or the optimization part?

                                               )
solver_stats = trunk(lsr1_LeNet; callback = callback_lsr1, max_time)
```

> 

Reviewer: Add a sentence there, to explain that the same example, using Knet and KnetNLPModels.jl, is available at this link [training documentation](https://jso.dev/KnetNLPModels.jl/stable/LeNet_Training/)

Author: do you want me to do this or the other comment fixed this ?

                                               )
solver_stats = trunk(lsr1_LeNet; callback = callback_lsr1, max_time)
```

> For more details on KnetNLPModels, we refer the reader to the KnetNLPModels [documentation](https://jso.dev/KnetNLPModels.jl/stable/)

Reviewer: ```suggestion
The same example exists also for KnetNLPModels.jl, and we refer the reader to the [LeNet training](https://jso.dev/KnetNLPModels.jl/stable/LeNet_Training/) documentation for more details.

```
